bitsandbytes/autograd/_functions.py

```diff
@@ -225,9 +226,10 @@ def forward(ctx, A, B, out=None, bias=None, state=MatmulLtState()):
         input_shape = A.shape
         if state.outlier_pool is None:
             state.outlier_pool = GlobalOutlierPooler.get_instance()
-        assert (
-            A.dtype == torch.float16
-        ), f"The input data type needs to be fp16 but {A.dtype} was found!"
+
+        # Cast A to fp16
+        A_dtype = A.dtype
+        A = A.to(torch.float16)
```

Comment:
This is exactly what we talked about, this should take care of bfloat16! One thing we need to think about is some kind of warning or something when this is first executed to make people aware that a cast happens and the operation quantization is performed in fp16.

Additionally, it would be great to have bf16 tests that verify everything works correctly with those inputs. I think you just need to change one [line](https://github.com/TimDettmers/bitsandbytes/blob/main/tests/test_autograd.py#L256) (and check if everything else is correctly executed).

Reply:
- added a warnings.warn the first time A is cast to a different dtype
- added FP16 tests; fine-tuned typecasts to fit into *almost* all thresholds
  - increased atol 0.0175 -> 0.02 in one specific case for BF16 (this does not affect FP16)
  - e.g. grad_bias is now computed on natural grad_output before it is cast (and before it loses precision)